<a href="https://colab.research.google.com/github/michiWS1920/nfl_data/blob/michi/NFL_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Introduction**













## **Main Objective of this Project**

<p align= "justify">Any strategic and mathematically optimizable game can profit from the use of analyzing historical data to improve decision quality in the future. Baseball was one of the first sports to recognize the value in broad datasets to optimize game-plans as seen in the film "Moneyball". 
In recent years data science has made its way into the National Football League (NFL) and is being used by teams to prepare for their opponents, as well as improving their training sessions and playcalling during the games. 


Since strategy driven sports like football are incorporating more and more technology into their matches/training, the acquired data is growing at a fast pace. These mentioned technologies used for data collection are motion sensors, as well as impact sensors on the football gear (e.g. helmet, shoulder pads, etc.) or more old fashioned ways of gathering historical play by play data manually. 

The desired outcome for this project is to find interesting and strategically valuable insights for the game of American Football. Specifically Playcalling, as well as the Passing and Rushing Game. <br>Visualizing these insights and drawing conclussions to improve decision quality, showing trends as well as building a predictive model are further objectives. </p>

## **Brief Explanation of American Football**

<p align= "justify">Since this project will go indepth on many strategic concepts of American Football it is necessary to explain the basics. Accompanying the analysis in this project will be more explanations to specific concepts and rules in order to clarify certain visualizations and findings. Unfortunately football is a rather complicated sport and the importance or magnitude of the analysis can only be understood if most strategic aspects are explained in advance.

American Football is a team sport played on a grass field with a leather ball, that can either be thrown or carried in the players' hands while running.  The field is 120 yards long and 53.3 yards wide, including both end zones at either end of the field. The game is played by two teams with 11 players on the field for each team simultaniously. A football game is 60 minutes long and is divided into 15-minute quarters.

A team has four chances (**"downs"**) to advance the ball 10 yards while the opposing team tries to stop it. <br>
This static division into offense and defense distinguishes American Football from other popular sports where possesion of the ball changes more quickly and leads to a dynamic switch in roles of offense and defense.<br> The mentioned downs can be seen as individual strategic attempts to progress down the field with the ball. After the ball hits the ground or a player gets tackled with the ball in posession the specific down is completed. Both teams line up in their chosen formations, where the previous down was completed. This regrouping, as well as the strategic break between every play enhances the importance of choosing a play according to the field position, formations of both teams and distance to the goal line. <br> If the ball is not advanced by 10 yards after four attempts posession of the ball will switch from one team to the other, as well as reverse the roles of offense and defense. If the ball is advanced 10 yards or more, the attacking team is rewarded with four new downs. </p>



###  **Scoring**

- Scoring is done by entering the opposing teams **"end zone"**, which is located at the end of either side of the field. When a player has possession of the ball and crosses the opposing teams goal line he scores a **"touchdown"**. A successful touchdown awards the scoring team with 6 points.
 - Immediatly after scoring a touchdown the scoring team can attempt to kick the ball from a set point through the two goal posts for 1 extra point.
 - The team can also run or pass the ball from a set point to the end zone with only one opportunity to score 2 additional points.
- Another option to score is the "field goal". A field goal is scored when the ball is kicked between the two goal posts located behind the opponent's end zone without a prior touchdown. A sucessful field goal awards the scoring team with 3 points.



>Whoever scores more points at the end of the 60 minute game-period wins the match. 

### **Positions**


As mentioned the teams are split into offense and defense.<br> Following will be two visualizations including the specific description of each individual position with a short summary of their tasks and objectives in the game.

**Offense**

<figure>
  <img src = "https://i.imgur.com/cYyOSiP.jpg" alt="Trulli" style="width:100%">
  <figcaption align = "left"> Fig. 1 - Offense Positions  </figcaption>
</figure>



*   **Quarterback (QB)**
  * The Quarterback is the most important player on the offensive side and most often the leader of the whole team. His objective is to throw the ball to his teammates or run with it to advance across the field and into the end zone of the defending team
*   **Running Back (RB)**
  * The Running Backs are usually handed the ball directly from the QB and are specialized in running with the ball in hand
* **Wide Receiver (WR)**
  * The Wide Receivers are lined up towards the sidelines of the field and are the primary throwing targets of the QB
* **Offensive Line (OL)**
  * The Offensive Linemen are lined up directly at the line of scrimmage where the play begins. Their primary concern is to stop the defense in their efforts to get to the QB

---

**Defense**

<figure>
  <img src = "https://i.imgur.com/KfdM6bY.jpg" alt="Trulli" style="width:100%">
  <figcaption align = "left"> Fig. 1 - Defense Positions  </figcaption>
</figure>

* **Defensive Line (DL)**
  * The Defensive Line are lined up directly at the line of scrimmage where the play begins. Their objective is to get to the opposing teams' QB/RB and tackle them to stop forward progress

* **Cornerback (CB)**
  * The Cornerbacks are responsible for the opposing teams Wide Receivers. Their objective is to tackle or intercept any balls thrown to the WRs.

* **Defensive Backfield (DB)**
  * The Denfesive Backfield is often used in a flexible manner. Players in the backfield can be used as reinforcement for the Defensive Line or to help the Cornerbacks.


## Data Sources


Until very recently the NFL provided a vast amount of data for free via their own API. Unfortunately this access is no longer available.
Because of this inconvenience we were not able to use the "[nflscrapR](https://github.com/maksimhorowitz/nflscrapR)" package, which is a community build wrapper for the NFL API. <br>Altough we could still use the raw data they have scraped in the past.

* This data includes csv [datasets](https://github.com/ryurko/nflscrapR-data) for play by play, roster and game data for the seasons from 2009 until 2019. This main data source for our project will be explored in the following basic EDA.


* Another data source for college football as well as complementary draft data is [Pro Football Refrence](https://www.pro-football-reference.com/). This website makes it possible to export player, draft and combine stats directly into csv format without the need of scraping.

* The final data source is [Next-Gen-Scrapy](https://github.com/ArrowheadAnalytics/next-gen-scrapy-2.0). This dataset was extracted using a very impressive scraper which translates pixels of an image into (x,y) coordinates. (Credit to [Sarah Mallepalle](https://arxiv.org/abs/1906.03339)) 
> "The pass charts are scraped from the NFL's official Next Gen Stats website, https://nextgenstats.nfl.com/charts/list/pass. After obtaining all of the pass chart images from the website, for every pass chart, the field is then undistorted, and all pass locations on the field relative to the line of scrimmage are extracted in (x,y) coordinate format."

# Setup

* Some packages included in tidyverse/ggplot are installed seperatly because of complications/errors that were observed during the project

In [ ]:
install.packages("ggrepel")
install.packages("scales")
install.packages("farver")
install.packages("ggExtra")
install.packages("hexbin")
install.packages("ggforce")

In [1]:
library(tidyverse)
library(ggplot2)
library(ggrepel)
library(ggforce)
library(scales)
library(farver)
library(ggExtra)
library(hexbin)

── Attaching packages ─────────────────────────────────────── tidyverse 1.3.0 ──

✔ ggplot2 3.3.2     ✔ purrr   0.3.4
✔ tibble  3.0.3     ✔ dplyr   1.0.2
✔ tidyr   1.1.1     ✔ stringr 1.4.0
✔ readr   1.3.1     ✔ forcats 0.5.0

── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()



ERROR: ignored

# Basic EDA nflscrapR Dataset

* Scraping the data and joining for multiple seasons
  * This includes Play by Play Data as well as Roster data for the regular seasons from 2009-2019
  * Some columns have to be removed since they are inconsistent over the seasons

This scraping can sometimes be buggy. Scraped data provided via googledrive

In [ ]:
first <- 2009 
last <- 2019 

datalist = list()
for (yr in first:last) {
    pbp <- read_csv(url(paste0("https://github.com/ryurko/nflscrapR-data/raw/master/play_by_play_data/regular_season/reg_pbp_", yr, ".csv"))) 
    games <- read_csv(url(paste0("https://raw.githubusercontent.com/ryurko/nflscrapR-data/master/games_data/regular_season/reg_games_", yr, ".csv"))) 
    pbp <- pbp %>% 
      inner_join(games %>% distinct(game_id, week, season)) %>% 
      select(-fumble_recovery_2_yards, -blocked_player_id, -fumble_recovery_2_player_id) 
    datalist[[yr]] <- pbp 
}

pbp_all <- dplyr::bind_rows(datalist)

* Some teams have changed their names and abbreviations over the seasons. Change teamnames for those to have consistent data

In [ ]:
pbp_all <- pbp_all %>% 
	mutate_at(vars(home_team, away_team, posteam, defteam, penalty_team), funs(case_when(
            . %in% "JAX" ~ "JAC",
            . %in% "STL" ~ "LA",
            . %in% "SD" ~ "LAC",
            TRUE ~ .
        ))) 

* Load already scraped data if scraping does not work (Error 400)

In [2]:
pbp_all <- readRDS("NFLdata.rds")

* Data Structure

In [11]:
str(pbp_all)

tibble [498,393 × 255] (S3: tbl_df/tbl/data.frame)
 $ play_id                             : num [1:498393] 46 68 92 113 139 162 183 207 228 253 ...
 $ game_id                             : num [1:498393] 2.01e+09 2.01e+09 2.01e+09 2.01e+09 2.01e+09 ...
 $ home_team                           : chr [1:498393] "PIT" "PIT" "PIT" "PIT" ...
 $ away_team                           : chr [1:498393] "TEN" "TEN" "TEN" "TEN" ...
 $ posteam                             : chr [1:498393] "PIT" "PIT" "PIT" "PIT" ...
 $ posteam_type                        : chr [1:498393] "home" "home" "home" "home" ...
 $ defteam                             : chr [1:498393] "TEN" "TEN" "TEN" "TEN" ...
 $ side_of_field                       : chr [1:498393] "TEN" "PIT" "PIT" "PIT" ...
 $ yardline_100                        : num [1:498393] 30 58 53 56 56 98 98 94 96 43 ...
 $ game_date                           : Date[1:498393], format: "2009-09-10" "2009-09-10" ...
 $ quarter_seconds_remaining           : num [1:498393

In [18]:
dim(pbp_all)

[1] 498393    255

* Huge dataframe with 498,393 individual plays over 10 seasons of football in the NFL and 255 variables to describe each play

* Note that many variables include **NA values** since they are specific to a certain event/play_type e.g. `run_location` will be NA unless the play was a rush



---



* All of the variables

In [13]:
names(pbp_all)

[1] "play_id"                             
  [2] "game_id"                             
  [3] "home_team"                           
  [4] "away_team"                           
  [5] "posteam"                             
  [6] "posteam_type"                        
  [7] "defteam"                             
  [8] "side_of_field"                       
  [9] "yardline_100"                        
 [10] "game_date"                           
 [11] "quarter_seconds_remaining"           
 [12] "half_seconds_remaining"              
 [13] "game_seconds_remaining"              
 [14] "game_half"                           
 [15] "quarter_end"                         
 [16] "drive"                               
 [17] "sp"                                  
 [18] "qtr"                                 
 [19] "down"                                
 [20] "goal_to_go"                          
 [21] "time"                                
 [22] "yrdln"                               
 [23] "ydstogo"                             
 [24] "ydsnet"                              
 [25] "desc"                                
 [26] "play_type"                           
 [27] "yards_gained"                        
 [28] "shotgun"                             
 [29] "no_huddle"                           
 [30] "qb_dropback"                         
 [31] "qb_kneel"                            
 [32] "qb_spike"                            
 [33] "qb_scramble"                         
 [34] "pass_length"                         
 [35] "pass_location"                       
 [36] "air_yards"                           
 [37] "yards_after_catch"                   
 [38] "run_location"                        
 [39] "run_gap"                             
 [40] "field_goal_result"                   
 [41] "kick_distance"                       
 [42] "extra_point_result"                  
 [43] "two_point_conv_result"               
 [44] "home_timeouts_remaining"             
 [45] "away_timeouts_remaining"             
 [46] "timeout"                             
 [47] "timeout_team"                        
 [48] "td_team"                             
 [49] "posteam_timeouts_remaining"          
 [50] "defteam_timeouts_remaining"          
 [51] "total_home_score"                    
 [52] "total_away_score"                    
 [53] "posteam_score"                       
 [54] "defteam_score"                       
 [55] "score_differential"                  
 [56] "posteam_score_post"                  
 [57] "defteam_score_post"                  
 [58] "score_differential_post"             
 [59] "no_score_prob"                       
 [60] "opp_fg_prob"                         
 [61] "opp_safety_prob"                     
 [62] "opp_td_prob"                         
 [63] "fg_prob"                             
 [64] "safety_prob"                         
 [65] "td_prob"                             
 [66] "extra_point_prob"                    
 [67] "two_point_conversion_prob"           
 [68] "ep"                                  
 [69] "epa"                                 
 [70] "total_home_epa"                      
 [71] "total_away_epa"                      
 [72] "total_home_rush_epa"                 
 [73] "total_away_rush_epa"                 
 [74] "total_home_pass_epa"                 
 [75] "total_away_pass_epa"                 
 [76] "air_epa"                             
 [77] "yac_epa"                             
 [78] "comp_air_epa"                        
 [79] "comp_yac_epa"                        
 [80] "total_home_comp_air_epa"             
 [81] "total_away_comp_air_epa"             
 [82] "total_home_comp_yac_epa"             
 [83] "total_away_comp_yac_epa"             
 [84] "total_home_raw_air_epa"              
 [85] "total_away_raw_air_epa"              
 [86] "total_home_raw_yac_epa"              
 [87] "total_away_raw_yac_epa"              
 [88] "wp"                                  
 [89] "def_wp"                            



---



Following will be an explanation of the most important variables in the dataframe



* `desc` = Detailed string description of the given play
* `game_id` = Ten digit identifier for NFL Game.
* `home_team` = String abbreviation for the home team
* `away_team` = String abbreviation for the away team
* `week` = Week in the season, between 1 and 16
* `posteam` = String abbreviation for the team in possesion of the football
* `defteam` = String abbreviation for the team on defense
* `yardline_100` = Distance in number of yards from the opponent's endzone for the posteam
* `down` = The down (1-4) for the given play
* `ydstogo` = Numeric yards distance to the next first down
* `play_type` = String indicating the type of play: pass if ball was thrown, run if rusher ran with the ball etc.
* `pass_attempt` = Binary indicator if the play was a pass
* `rush_attempt` = Binary indicator if the play was a run
* `complete_pass` = Binary indicator if the pass was completed
* `touchdown` = Binary indicator if the play resulted in a touchdown
* `yards_gained` = Numeric yards gained (or lost) by the possessing team
* `pass_location` = String indicator for pass location: left, middle or right
* `air_yards` = Numeric value for distance in yards from the line of scrimmage at where the receiver caught the ball or did not catch it
* `total_home_score` = Score for the home team at the start of the play
* `total_away_score` = Score the away team at the start of the play
* `score_differential` = Score differential between posteam and defteam at the start of the play
* `penalty` = Binary indicator for whether or not a penalty occured
* `penalty_team` = String abbreviation of the team with the penalty
* `penalty_yards` = Yards gained (or lost) by the posteam from the penalty




Lets look at an example - some of the variables in use.

In [12]:
pbp_all %>%
select(season, week, home_team, away_team, total_home_score, total_away_score, down, ydstogo, play_type, yards_gained, penalty) %>%
sample_n(5)

season week home_team away_team total_home_score total_away_score down
1 2012    3   WAS       CIN        6               7                2   
2 2019   16   ATL       JAC       14               0                1   
3 2010    3   NE        BUF        7               3                1   
4 2019   16   PHI       DAL       10               6                2   
5 2016   11   CAR       NO         3               3                2   
  ydstogo play_type yards_gained penalty
1 12      pass      22           0      
2 10      pass      16           0      
3 10      run        3           0      
4  9      pass       8           0      
5 10      run       -2           0

Following are more complicated but important variables

* `ep` = **Expected Points (EP)** by the posteam before the play.  Is calculated by value of the current down, distance, and field position situation in terms of future expected net point advantage. Examples: First down with one yard to go to the endzone will represent an EP near 6 (since a touchdown yields 6 points)


* `epa` = **Expected Points Added (EPA)** by the posteam at the end of a play. Can be seen as the difference between Expected Points (EP) at the start of the play and the EP at the end of the play. 


* `wp` = **Win Probability (WP)**. The probability that a team will win a game, given a particular combination of circumstances including score, time remaining, field position, down and to go distance. WP is based on a model built on actual outcomes of NFL games that featured similar circumstances.


* `wpa` = **Win Probability Added (WPA)**. Difference between a team's Win Probability at the start of a play and the WP at the end of the play. WPA can be seen as the measure of a play's impact on the outcome of a game



---


* Showcasing some of the more complicated variables (`epa, ep, wpa, wp`) as well as the `desc` variable to illustrate how detailed the written out description is

In [17]:
pbp_all %>%
filter(touchdown == 1) %>%
select(desc, epa, ep, wpa, wp) %>%
tail()

desc                                                                                                                                                                  
1 (:32) D.Samuel left end for 30 yards, TOUCHDOWN. The Replay Official reviewed the runner was inbounds ruling, and the play was Upheld. The ruling on the field stands.
2 (5:53) (Shotgun) R.Wilson pass short left to T.Lockett for 14 yards, TOUCHDOWN.                                                                                       
3 (3:28) R.Mostert right tackle for 2 yards, TOUCHDOWN.                                                                                                                 
4 (9:59) J.Jones reported in as eligible.  M.Lynch up the middle for 1 yard, TOUCHDOWN.                                                                                 
5 (5:55) (Shotgun) R.Mostert left guard for 13 yards, TOUCHDOWN.                                                                                                        
6 (3:40) (Shotgun) R.Wilson pass short left to DK.Metcalf for 14 yards, TOUCHDOWN.                                                                                      
  epa      ep       wpa        wp       
1 3.068499 3.931501 0.08998383 0.7045027
2 3.535852 3.464148 0.09913199 0.1709520
3 1.159430 5.840570 0.01916091 0.8613702
4 0.530851 6.469149 0.01723658 0.2586973
5 2.630784 4.369216 0.04692250 0.8801631
6 3.225973 3.774027 0.10801224 0.1003556

* The `desc` column will be very useful. With the use of the **stringr** package we can extract additional information!



---


The dataset is missing some crucial information that would be valuable for further analysis like:
* Who was the Passer / Rusher / Receiver in the play

Further Clean up process:
* For `play_type == "no_play"`, yards_gained column has to be changed to NA since these plays were invalid because of a foul or other disturbance on the play.

* Filter out plays that have NA values considering `EPA` and `Posteam`, because these are useless and probably faulty to begin with.

* Filter out multiple `play_types` - "*punt*", "*field_goal", "kickoff"*, "*extra_point", "qb_kneel"* and "*qb_spike*" since we are not interested in these for now

Credit to [Ben Baldwin](https://gist.github.com/guga31bb/5634562c5a2a7b1e9961ac9b6c568701) for these pre-processing steps (specifically the stringr part).

In [13]:
pbp_all_rp <- pbp_all %>%
      filter(!is.na(epa), !is.na(posteam), play_type=="no_play" | play_type=="pass" | play_type=="run") %>% 
	mutate( 
	pass = if_else(str_detect(desc, "( pass)|(sacked)|(scramble)"), 1, 0), 
	rush = if_else(str_detect(desc, "(left end)|(left tackle)|(left guard)|(up the middle)|(right guard)|(right tackle)|(right end)") & pass == 0, 1, 0),
	success = ifelse(epa>0, 1 , 0),
	passer_player_name = ifelse(play_type == "no_play" & pass == 1,
              str_extract(desc, "(?<=\\s)[A-Z][a-z]*\\.\\s?[A-Z][A-z]+(\\s(I{2,3})|(IV))?(?=\\s((pass)|(sack)|(scramble)))"),
              passer_player_name),
        receiver_player_name = ifelse(play_type == "no_play" & str_detect(desc, "pass"), 
              str_extract(desc, "(?<=to\\s)[A-Z][a-z]*\\.\\s?[A-Z][A-z]+(\\s(I{2,3})|(IV))?"),
              receiver_player_name),
        rusher_player_name = ifelse(play_type == "no_play" & rush == 1, 
              str_extract(desc, "(?<=\\s)[A-Z][a-z]*\\.\\s?[A-Z][A-z]+(\\s(I{2,3})|(IV))?(?=\\s((left end)|(left tackle)|(left guard)|(up the middle)|(right guard)|(right tackle)|(right end)))"),
              rusher_player_name),
	name = ifelse(!is.na(passer_player_name), passer_player_name, rusher_player_name),
	yards_gained=ifelse(play_type=="no_play",NA,yards_gained), play=1
	) 

In [15]:
dim(pbp_all_rp)

[1] 399162    260

Through the filtering and clean up process the play count is now about 100,000 lower and the variable count increased by 5.